### PR TITLE
[#IC 278] Handler HTTP

### PR DESCRIPTION
**Pull request by LoFrance**
HTTP Handler:

- receives an HTTP request in IN
  - The type of the request will contain the **OrganizationFiscalCode** and the **SourceId**
  - OrganizationFiscalCode is needed to retrieve the APIM Account to retrieve the targetId (this step is done by another Handler - not this one!!!)
  - Update all subscriptions (status! == completed) belongs to the OrganizationFiscalCode and SourceId with status at pending
- writes in OUT a message on the queue **organizationToMigrate**:
   - a message containing the targetId, organizationFiscalCode and sourceId
- writes in OUT RES a response 202 or a response 500 if I have a status update error on PostgreSQL

----
Depends on https://github.com/pagopa/io-infra/pull/117

### CODE REVIEW

File: ClaimOwnership/index.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import { getConfigOrThrow } from "../utils/config";
+import getPool from "../utils/db";
+import ClaimOwnershipHandler from "./handler";
+
+const config = getConfigOrThrow();
+const client = getPool(config);
+const setupExpress = (): express.Express => {
+  const app = express();
+  app.post(
+    "/api/v1/organizations/:organizationFiscalCode/:delegate_id",
```
Comment (balanza) [inline ClaimOwnership/index.ts:14]: Please review this path, I'm not sure it's correct

File: ClaimOwnership/handler.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import { queryDataTable } from "../utils/db";
+import { OrganizationQueueItem } from "./types";
+
+type Handler = (
+  context: Context,
+  organizationFiscalCode: OrganizationFiscalCode,
+  delegateId: NonEmptyString
+) => Promise<IResponseSuccessJson<void> | IResponseErrorInternal>;
+
+export const updateSqlSubscription = (dbConfig: IDecodableConfigPostgreSQL) => (
```
Comment (balanza) [inline ClaimOwnership/handler.ts:40]: I think the function name is misleading: it let me think it updates subscription data; instead, I see it _sets migration status to PENDING for **every** subscription of a Delegate for an Organization_.

(also, a comment might be helpful for the reader to understand what to expect without digging into the function implementation details)

File: ClaimOwnership/handler.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    .where({ organizationFiscalCode })
+    .where({ sourceId })
+    .where("status", "!=", SubscriptionStatus.COMPLETED)
+    .update({
+      status: SubscriptionStatus.PROCESSING
+    })
+    .from(dbConfig.DB_TABLE)
+    .toQuery() as NonEmptyString;
+
+export const updateSubscriptionsByOrganizationFiscalCodeAndSourceId = (
```
Comment (balanza) [inline ClaimOwnership/handler.ts:58]: Same as the comment above (misleading function name)

File: ClaimOwnership/types.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import {
+  NonEmptyString,
+  OrganizationFiscalCode
+} from "@pagopa/ts-commons/lib/strings";
+import * as t from "io-ts";
+
+// The shape of the item expected in the queue
+export type OrganizationQueueItem = t.TypeOf<typeof OrganizationQueueItem>;
```
Comment (balanza) [inline ClaimOwnership/types.ts:8]: Naming: imho `OrganizationQueueItem` doesn't communicate enough what the item is for. Isn't that a `Claim` "entity" (defined by the tuple Organization-Delegate)?

File: ClaimOwnership/handler.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    TE.mapLeft(flow(toPostgreSQLErrorMessage, toPostgreSQLError))
+  );
+
+export const organizationMessageToQueue = (context: Context) => (
+  organizationFiscalCode: OrganizationFiscalCode,
+  sourceId: NonEmptyString
+): E.Either<boolean, boolean> =>
+  pipe(
+    { organizationFiscalCode, sourceId },
+    OrganizationQueueItem.decode,
```
Comment (balanza) [inline ClaimOwnership/handler.ts:77]: Maybe this is the case for encoding, not decoding
* encoding --> from our domain to the outside world
* decoding --> from the outside world to our domain

Please not that this simplifies a lot the process (no error flow is intended)

File: ClaimOwnership/handler.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  organizationFiscalCode: OrganizationFiscalCode,
+  sourceId: NonEmptyString
+): TE.TaskEither<IDbError, QueryResult<QueryResultRow>> =>
+  pipe(
+    updateSqlSubscription(config)(organizationFiscalCode, sourceId),
+    sql => queryDataTable(connect, sql),
+    TE.mapLeft(flow(toPostgreSQLErrorMessage, toPostgreSQLError))
+  );
+
+export const organizationMessageToQueue = (context: Context) => (
```
Comment (balanza) [inline ClaimOwnership/handler.ts:71]: Naming: what's this function for? Is it clear that it has a side effect (it writes to the queue)? It makes me think it is just a conversion from a structure X to Y.

Please decide whether:
* to make the function pure, thus moving the side effect outside
* to keep the function to encapsulate side effects, and then it needs to have an explicit name

**Review by balanza — CHANGES_REQUESTED** [review]
Implementation is good, however I found it difficult to follow the underneath reasoning because of misleading and obscure naming choices. 

BTW why draft?

File: ClaimOwnership/handler.ts (hunk 2 above)
Comment (LoFrance) [reply ClaimOwnership/handler.ts]: You're right, it's done by f129126

File: ClaimOwnership/types.ts (hunk 4 above)
Comment (LoFrance) [reply ClaimOwnership/types.ts]: Refactor with 79f7545

File: ClaimOwnership/handler.ts (hunk 6 above)
Comment (LoFrance) [reply ClaimOwnership/handler.ts]: The function now is splitted in 2 easier function, one to create the message and one to write the message to the Queue. Maybe it's better in this way: d839eb9 ?

File: ClaimOwnership/handler.ts (hunk 5 above)
Comment (LoFrance) [reply ClaimOwnership/handler.ts]: It's already done, I think we have talk about it in a previous meeting.

File: ClaimOwnership/index.ts (hunk 1 above)
Comment (LoFrance) [reply ClaimOwnership/index.ts]: Yep, it's fix by a4cb626

**Review by balanza — APPROVED** [review]

**Pull request merged**